[CORE 8273] Adjust license warning text to include trial license links

BenPope opened the pull request:
Adjust license warning text to include trial license links.

Adjust CreateACLs response error to `invalid_config` from `unknown_server_error`, to match topic create/alter.

## Backports Required

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

* none

oleiman commented on src/v/features/enterprise_feature_messages.h:
```diff
@@ -0,0 +1,119 @@
+inline ss::sstring cluster_property(std::string_view name, auto value) {
+    return ssx::sformat(
+      fmt::runtime(
+        "Rejected {}: '{}' '{}' is restricted to clusters with an Enterprise "
```
nitpick: maybe (just maybe):

```suggestion
        "Rejected {}: '{}' - '{}' is restricted to clusters with an Enterprise "
```

BenPope replied:
That's not what the requirements say. @mattschumpert ?

oleiman replied:
Oh, there are reqs...ignore me!

oleiman commented on src/v/features/enterprise_feature_messages.h:
```diff
@@ -0,0 +1,119 @@
+
+inline ss::sstring default_leaders_preference() {
+    return ssx::sformat(
+      R"({} to use the enterprise feature "leadership pinning".  This feature is disabled.  The values of the cluster property <value> and the topic property "redpanda.leaders.preference" are being ignored. {})",
```
> `<value>`

Should that contain something specific?

BenPope replied:
Not according to the requirements.

IoannisRP replied:
For this specific one we didn't include the values in the previous version. The issue is that to report all the potential values that might be in violation you have the enterprise property and all the topics that might have set their property to anything not-default.
Printing all the topics that are in violation could (theoretically) lead to overallocation warning.
One option was to limit the number of topic names being printed, but that required a bit of extra logic.
In the end it was decided to not print any of the values for simplicity.

oleiman commented on src/v/features/enterprise_feature_messages.h:
```diff
@@ -0,0 +1,119 @@
+
+inline ss::sstring topic_property(auto feature) {
+    return ssx::sformat(
+      "{} enable {}. {}.", required, feature, request_or_trial);
```
```suggestion
      "{} to enable {}. {}.", required, feature, request_or_trial);
```

oleiman commented on src/v/cluster/config_manager.cc:
```diff
@@ -658,6 +659,8 @@ apply_local(const cluster_config_delta_cmd_data& data, bool silent) {
 
             auto validation_err = property.validate(val);
             if (validation_err.has_value()) {
+                auto msg = features::enterprise_error_message::cluster_property(
+                  u.key, val);
```
this looks new. I'm surprised we print an enterprise error on a straight validation error. am I missing something?

oleiman commented on src/v/cluster/partition_balancer_backend.cc:
```diff
@@ -370,9 +371,9 @@ ss::future<> partition_balancer_backend::do_tick() {
-          "A Redpanda Enterprise Edition license is required to use enterprise "
-          "feature \"partition_autobalancing_mode\" with value \"{}\". "
-          "Behavior is being restricted to \"{}\".",
+          "()",
```
```suggestion
          "{}",
```

oleiman reviewed (commented):
few copy edits, otherwise looks awesome

IoannisRP commented on tests/rptest/tests/enterprise_features_license_test.py:
```diff
@@ -64,6 +64,12 @@ class Feature(IntEnum):
     Feature.fips  # NOTE(oren): omit because it's too much of a pain for CDT
 ]
 
+ENTERPRISE_LICENSE_NAG = "A Redpanda Enterprise Edition license is required"
+
+
+def has_license_nag(redpanda):
+    return redpanda.search_log_any(ENTERPRISE_LICENSE_NAG)
+
```
this feels like it would be better suited to be a member of redpanda (or somewhere?).
If feels wrong to import functionality from anther test.

IoannisRP reviewed (commented):
Very nice groupings!
just a couple of minor comments

BenPope commented on src/v/cluster/config_manager.cc:
```diff
@@ -658,6 +659,8 @@ apply_local(const cluster_config_delta_cmd_data& data, bool silent) {
 
             auto validation_err = property.validate(val);
             if (validation_err.has_value()) {
+                auto msg = features::enterprise_error_message::cluster_property(
+                  u.key, val);
```
Not sure how it got there!

vbotbuildovich commented:
<!--NEW FAILURES-->
<!--53bd34bd-beda-4fb5-9ee3-69403b392bd5-->
<!--body-->
non flaky failures in https://buildkite.com/redpanda/redpanda/builds/58582#01935508-57f3-4541-b166-2b90a9d45266:
```
"rptest.tests.rbac_test.RBACLicenseTest.test_license_nag"
"rptest.tests.redpanda_kerberos_test.RedpandaKerberosLicenseTest.test_license_nag"
```
<!--/body-->

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/58582-->
<!--a1376b50-55d6-4e3e-afde-48f35254a18a-->
<!--title-->
## Retry command for [Build#58582](https://buildkite.com/redpanda/redpanda/builds/58582)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/rbac_test.py::RBACLicenseTest.test_license_nag
tests/rptest/tests/redpanda_kerberos_test.py::RedpandaKerberosLicenseTest.test_license_nag
```
<!--/body-->

BenPope commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/b4209a487f22cadd01316ae03dc49bdd72545648..0f271744f477b0f082277c67b7050cbd958604fc)
* Improve some formatting strings
* Switch license strings for admin

oleiman reviewed (approved):
lgtm

BenPope commented:
CI Failures:
* The bazel failure appears to relate to https://github.com/redpanda-data/redpanda/pull/24226 - [CI failed on it's merge](https://buildkite.com/redpanda/redpanda/builds/58583#019354cd-e4b9-4d48-bd5f-f01d97e20596)

vbotbuildovich commented:
<!--UNIT FAILURES-->
<!--7edbc549-c0ae-427b-b9fb-40e403f4fe7e-->
<!--body-->
the below tests from https://buildkite.com/redpanda/redpanda/builds/58599#0193558e-8be9-460e-ad15-7de7661242ba have failed and will be retried
```
gtest_raft_rpunit
```
<!--/body-->

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/58599-->
<!--6077c933-b568-434c-aa92-e68f4cc9607a-->
<!--title-->
## Retry command for [Build#58599](https://buildkite.com/redpanda/redpanda/builds/58599)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/datalake/coordinator_retention_test.py::CoordinatorRetentionTest.test_retention@{"cloud_storage_type":1}
```
<!--/body-->

vbotbuildovich commented:
<!--RETRIED FAILURES-->
<!--67bcaa23-46dd-4143-aae4-89a2ffbae86f-->
<!--body-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/58599#019355d4-876a-4024-b029-41989a7c63db
<!--/body-->

BenPope commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/edb62f095f22f0b76f458b89c4346034719cf6c8..78a0d9d8fe5cf2cd78fc99a456381395c893269d)
* Moved `has_license_nag` to redpanda
* Removed `fmt::runtime`
* Changed CreateACL response`error_code` from `unknown_server_error` to `invalid_config`

IoannisRP reviewed (approved):
lgtm

vbotbuildovich commented:
/backport v24.3.x

oleiman commented on src/v/kafka/server/server.cc:
```diff
@@ -1578,13 +1579,13 @@ ss::future<response_ptr> create_acls_handler::handle(
-                creatable_acl_result{.error_code = ec});
               if (results[i] == cluster::errc::feature_disabled) {
-                  response.data.results.back().error_message.emplace(
-                    "An enterprise license is required to create an ACL with a "
-                    "role binding");
+                  response.data.results.emplace_back(
+                    error_code::invalid_config,
+                    features::enterprise_error_message::acl_with_rbac());
+              } else {
+                  response.data.results.emplace_back(
+                    map_topic_error_code(results[i]));
```
nice

oleiman reviewed (commented):
lgtm
